[Grid] Start considering overflow position from fallback alignment of distributed alignment values.

sammygill opened the pull request:
#### 9aa15d360c9652bf887c6c33776a81fb37010356
<pre>
[Grid] Start considering overflow position from fallback alignment of distributed alignment values.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278118">https://bugs.webkit.org/show_bug.cgi?id=278118</a>
<a href="https://rdar.apple.com/133865365">rdar://133865365</a>

Reviewed by Alan Baradlay.

We currently apply alignment safety when the content would overflow and
our content alignment value has an overflow position of safe. This will
only work if safe was specified alongside the content position, but
we also need to consider the fallback alignment of the content
distribution values if that was used instead. Certain content
distribution values have a fallback alignment of safe so we will need
to respect those.

In order to do this we need to change resolveContentDistributionFallback
to return a pair OverflowAlignment and ContentPosition. We will then
consider this OverflowAlignment value when checking for alignment safety.

I also moved creating the &quot;position,&quot; now &quot;usedContentPosition&quot; to right
before performing the &lt;content-position&gt; / fallback positioning since
that is when we will need it. If the ContentDistribution value is not
default then we should not need to also check if ContentPosition is
normal since that should already be the case.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-place-content-001-expected.txt:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::resolveContentDistributionFallback):
(WebCore::RenderGrid::computeContentPositionAndDistributionOffset):

Canonical link: <a href="https://commits.webkit.org/282302@main">https://commits.webkit.org/282302@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8019a2b84cbd1921a0e520577ef4d082f428483a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62733 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66753 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13337 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13621 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50613 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9210 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39147 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54350 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31292 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35855 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11676 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57404 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12007 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68448 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6679 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57935 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58123 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13923 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5591 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37909 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38989 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40100 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2305,23 +2305,23 @@ void RenderGrid::gridAreaPositionForGridItem(const RenderBox& gridItem, GridTrac
-        return ContentPosition::Start;
+        return { OverflowAlignment::Default, ContentPosition::Start };
     case ContentDistribution::SpaceAround:
-        return ContentPosition::Center;
+        return { OverflowAlignment::Safe, ContentPosition::Center };
     case ContentDistribution::SpaceEvenly:
-        return ContentPosition::Center;
+        return { OverflowAlignment::Safe, ContentPosition::Center };
     case ContentDistribution::Stretch:
-        return ContentPosition::Start;
+        return { OverflowAlignment::Default, ContentPosition::Start };
     case ContentDistribution::Default:
-        return ContentPosition::Normal;
+        return { OverflowAlignment::Default, ContentPosition::Normal };
```
Since start is functionally safe, is a pair here actually necessary since every possible input value returns safe?

Also maybe you could just use `StyleContentAlignmentData::isCentered()` instead of having this function?

fantasai replied:
Also, if it makes it easier in the subsequent code, maybe StyleContentAlignmentData can just return safe explicitly for all the ContentDistribution values.

sammygill replied:
Yeah that makes sense but I think I slightly prefer this approach since it makes it much more clear what is going on around the call site and I feel like just using `isCentered` would slightly detract at that. 

In theory if we ever add a new `ContentDistribution` value we would also be forced to adjust this code properly.

sammygill committed:
[Grid] Start considering overflow position from fallback alignment of distributed alignment values.
https://bugs.webkit.org/show_bug.cgi?id=278118
rdar://133865365

Reviewed by Alan Baradlay.

We currently apply alignment safety when the content would overflow and
our content alignment value has an overflow position of safe. This will
only work if safe was specified alongside the content position, but
we also need to consider the fallback alignment of the content
distribution values if that was used instead. Certain content
distribution values have a fallback alignment of safe so we will need
to respect those.

In order to do this we need to change resolveContentDistributionFallback
to return a pair OverflowAlignment and ContentPosition. We will then
consider this OverflowAlignment value when checking for alignment safety.

I also moved creating the "position," now "usedContentPosition" to right
before performing the <content-position> / fallback positioning since
that is when we will need it. If the ContentDistribution value is not
default then we should not need to also check if ContentPosition is
normal since that should already be the case.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-place-content-001-expected.txt:
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::resolveContentDistributionFallback):
(WebCore::RenderGrid::computeContentPositionAndDistributionOffset):

Canonical link: https://commits.webkit.org/282302@main
